cranelift/fuzzgen/src/lib.rs

```diff
@@ -8,20 +8,55 @@ use cranelift::codegen::ir::Function;
 use cranelift::codegen::Context;
 use cranelift::prelude::*;
 use cranelift_native::builder_with_options;
+use std::fmt;
 
 mod config;
 mod function_generator;
 
 pub type TestCaseInput = Vec<DataValue>;
 
-#[derive(Debug)]
 pub struct TestCase {
     pub func: Function,
     /// Generate multiple test inputs for each test case.
     /// This allows us to get more coverage per compilation, which may be somewhat expensive.
     pub inputs: Vec<TestCaseInput>,
 }
 
+impl fmt::Debug for TestCase {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            r#";; Fuzzgen test case
+
+test interpret
+test run
+set enable_llvm_abi_extensions
+target aarch64
+target s390x
+target x86_64
+
+"#
+        )?;
+
+        writeln!(f, "{}", self.func)?;
+
+        for input in self.inputs.iter() {
+            let args = input
+                .iter()
+                .map(|val| format!("{}", val))
+                .collect::<Vec<_>>()
+                .join(", ");
+
+            // TODO: We don't know the expected outputs, maybe we can run the interpreter
+            // here to figure them out? Should work, however we need to be careful to catch
+            // panics in case its the interpreter that is failing.
+            writeln!(f, "; run: {}({}) == TODO", self.func.name, args)?;
```

Comment:
How about emitting a type-appropriate default value, like `0` or `0x0.0` or whatever? Then someone trying to use this output can just run it and see from the error messages what the current implementation returns. If you use "TEST" like this then running the generated test just returns an unhelpful parse failure message.

I'm happy to merge this as-is but I'll hold off in case you want to give that a try.

Fiddling around with this reveals some usability annoyances with `clif-util` too. It'd be nice if it reported what line number the error was on, and if it could optionally keep going after a failure to report all errors in the test.

Reply:
That sounds like a great idea!